### PR TITLE
Add new guide "Getting Started with PlatformIO"

### PR DESCRIPTION
**Pull request by valeros**
Hi there! First of all, thanks a lot for your work, it's a great project for developers who is just getting started with the STMicro ecosystem. 

This PR adds some background material on how to use the STM32 products with PlatformIO. The guide shows a complete development workflow, starting from creating a project up to launching a debug session to interact with the code.

cc @ivankravets

### CODE REVIEW

**Comment by ThomasGravekamp** [comment]
Hi @valeros!

Sorry for my late response. I appreciate the effort you put into writing this guide. I, however, fail to understand why you chose to create a PR here. A guide like this, in my opinion, would be a better fit on the PlatformIO documentation website. That is the first place I would go looking for a guide like this.

I want to focus documentation on the STM32-base website on these things:

- General guides and information on Arm Cortex-M cores
- General guides and information on STM32 microcontrollers and development boards
- Guides specific to STM32-base

I hate to say this, but I do not want to merge this PR. The reason is simple. The documentation for this project should not replace another project's (lack of) documentation. And I do not want to have the responsibility of maintaining this guide. I am not following the PlatformIO project to such a degree that I will be able to keep this guide up to date. I already have a hard time keeping the project in its current state up to date.

I would advise you to create a PR for this guide on PlatformIO's (documentation) repository. I do want to add a link to PlatformIO as one of the IDE's available for STM32 microcontrollers.

**Comment by valeros** [comment]
Hi @ThomasGravekamp ! Actually, I'm a member of PlatformIO team. The main reason for this PR is solely to provide a comprehensive overview of the PlatformIO tools, so newcomers to the STM32 ecosystem will be aware of alternative instruments. We don't have any doubts that maintaining an open source is hard and we appreciate your efforts, so don't worry about maintaining PlatformIO-related pages, it's totally our burden to keep the tutorial up to date.

**Comment by ThomasGravekamp** [comment]
Thanks for the clarification @valeros !

There is one thing I still do not understand. Why do you think the stm32-base documentation is better suited to have this guide than the PlatformIO docs? Do you have reasons to believe that people end up on on the stm32-base documentation, looking for guides/tutorials for PlatformIO?

The reason I set the scope of this documentation to not include guides on other projects and IDE's is to not having to maintain them. So you taking the responsibility to keep it up-to-date is a relieve. 

And there is one last thing. STM32-base's philosophy is to expose its users, or at least let them easily access, all facets of the toolchain which are needed to compile a program for STM32 microcontrollers. On the other hand there is PlatformIO, which in my experience, tries to make everything as simple as it can to set up and get users going. What do you think of that difference? Is that something you already knew about STM32-base?

**Comment by valeros** [comment]
> There is one thing I still do not understand. Why do you think the stm32-base documentation is better suited to have this guide than the PlatformIO docs? Do you have reasons to believe that people end up on on the stm32-base documentation, looking for guides/tutorials for PlatformIO?

What do you think if we move this tutorial to the PlatformIO docs and refactor this `Getting Started with PlatformIO` page to something more generic, possibly with links to other STM32-related tutorials and examples available within the PlatformIO ecosystem? This way the STM32-Base docs will remain to be a great starting point that contains all that scattered info all over the internet. Users won't be overwhelmed by a heavyweight tutorial and only those who is interested can proceed to the full-featured PlatformIO documentation. 

> And there is one last thing. STM32-base's philosophy is to expose its users, or at least let them easily access, all facets of the toolchain which are needed to compile a program for STM32 microcontrollers. On the other hand there is PlatformIO, which in my experience, tries to make everything as simple as it can to set up and get users going. What do you think of that difference? Is that something you already knew about STM32-base?

Personally, I don't see here any difference, only synergy. Both projects are focused on lowering the entry barriers for people who is interested in the STM32 MCUs. Why don't let people decide which way is more suitable for them?

**Comment by ThomasGravekamp** [comment]
> What do you think if we move this tutorial to the PlatformIO docs and refactor this `Getting Started with PlatformIO` page to something more generic, possibly with links to other STM32-related tutorials and examples available within the PlatformIO ecosystem?

I am not against hosting a `Getting Started with PlatformIO` guide when you take on the responsibility of keeping it up to date. What bothers me is that you seem to keep avoiding answering my question: Why do you think that STM32-base is a better place for hosting this guide than the PlatformIO docs?

I think I know the answer to that question. You said "so newcomers to the STM32 ecosystem will be aware of alternative instruments". If that is your true goal, a mention of PlatformIO in the existing guide `Getting started with STM32` would have been enough to meet that goal. It looks like you are looking for more exposure for the PlatformIO project in general. If this is true, I would like you to just say so and be honest about that.

> This way the STM32-Base docs will remain to be a great starting point that contains all that scattered info all over the internet.

On the other hand, I can certainly agree on this. This is one of the reasons I started this project.

> Users won't be overwhelmed by a heavyweight tutorial and only those who is interested can proceed to the full-featured PlatformIO documentation.

I am not concerned about users being overwhelmed. The `Getting Started with STM32-base` may even be a more heavy weight guide.

We might be able to make this work, under these terms:

- You, or another maintainer of PlatformIO, will be responsible for keeping this guide up to date
- Issues reported on this guide must be handled by you or another maintainer of PlatformIO
- We'll label this guide to make clear it is written by and maintained by PlatformIO
- I can and will remove the guide when I think it does not fit the STM32-base docs any longer. For example, when the guide is no longer maintained.

**Comment by valeros** [comment]
> It looks like you are looking for more exposure for the PlatformIO project in general. If this is true, I would like you to just say so and be honest about that.

Yes, it's not a secret. We are interested in attracting more developers and it seems that STM32-Base is a great place to present PlatformIO for a broader audience and a full-featured tutorial might help them quickly evaluate whether PlatformIO is worth the effort.

> We might be able to make this work, under these terms:

All of those terms sound perfectly reasonable. Do you have any expectations how that label should look like? For example:
```
Please note this tutorial is maintained by the PlatformIO team, please report any issues with PlatformIO directly to its developers.
```

**Comment by ThomasGravekamp** [comment]
> Yes, it's not a secret. We are interested in attracting more developers and it seems that STM32-Base is a great place to present PlatformIO for a broader audience and a full-featured tutorial might help them quickly evaluate whether PlatformIO is worth the effort.

Alright. Please remember to be upfront about these things in the future. I personally find it really annoying to have to dig so deep to uncover someone's true intentions. Things would have gone a lot faster if you've said this in your initial comment.

> All of those terms sound perfectly reasonable. Do you have any expectations how that label should look like?

No, not really. I will look into creating some sort of info banner in which we can put the message "Note: This guide is maintained by the PlatformIO team.". The guide will become part of this repo, so issues should be reported on this repo. I can handle the small things like spelling errors, like I already do for the other content. Those are usually one minute of work.

I will assign issues like "Step X does not work" to you. You can then decide to handle these issues yourself or to let someone else handle them for you.

I'll look into the actual content of your PR coming days.

File: guides/platformio.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+
+In the next step, we need to select `ST Nucleo-F401RE` as the development board, `STM32Cube` as the framework, and a path to the project location (or use the default one):
+
+![Getting Started with PlatformIO Project Init]({{ site.url }}/assets/img/guides/platformio/setting-up-project-2.png)
+{: style="text-align: center;"}
+
+##### Please note:
+
+Processing the selected project may take a considerable amount of time (PlatformIO will download and install all required packages) and after these steps, we have a fully configured project that is ready for developing code with the STM32Cube framework.
```
Comment (ThomasGravekamp) [inline guides/platformio.md:42]: `... amount of time as PlatformIO will download... ...all required packages. After these steps...`

File: guides/platformio.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+![Getting Started with PlatformIO Debug Initialization]({{ site.url }}/assets/img/guides/platformio/setting-up-project-7.png)
+{: style="text-align: center;"}
+
+We need to wait some time while PlatformIO is initializing a debug session and when the first line after the `main` function is highlighted we are ready to debug:
+
+![Getting Started with PlatformIO Debug Session]({{ site.url }}/assets/img/guides/platformio/setting-up-project-8.png)
+{: style="text-align: center;"}
+
+We can walk through the code using control buttons, set breakpoints, see peripheral registers, add variables to the `Watch` window:
```
Comment (ThomasGravekamp) [inline guides/platformio.md:121]: walk -> step

`...registers, and add variables...`

**Review by ThomasGravekamp — CHANGES_REQUESTED** [review]
As a general comment:

- Why have you not written this guide to be used with a Blue Pill board?
- Can you get rid of "we" throughout the guide? This is just a matter of taste, and for the docs that are already there I seem to have avoided using "we" in guides.

**Comment by valeros** [comment]
> Why have you not written this guide to be used with a Blue Pill board?

I just wanted to use an official ST board with an embedded debug probe and I happened to have Nucleo-F401RE at hand.

> Can you get rid of "we" throughout the guide? This is just a matter of taste, and for the docs that are already there I seem to have avoided using "we" in guides.

Done.

**Comment by ThomasGravekamp** [comment]
> I just wanted to use an official ST board with an embedded debug probe and I happened to have Nucleo-F401RE at hand.

Alright, fair enough. I already suspected that you wanted a board with built-in ST-link module.

I've not looked into the changes you've made. I'll try to do that coming days. One thing you can do is rebase on the master branch. You can then use this include `{% include note-info.html content="Note that this tutorial is written by and maintained by the PlatformIO team." %}` to show the info message above the guide.

File: guides/platformio.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+void SysTick_Handler(void) {
+  HAL_IncTick();
+}
+```
+
+After this step the project is ready for compiling and uploading.
+
+## Compiling and Uploading the Firmware
+
+Now the project can be built. The following options are available to compile the final firmware : Build option on the `Project Tasks` menu, Build button on PlatformIO Toolbar, using Command Palette `View: Command Palette > PlatformIO: Build`, using the Task Menu `Tasks: Run Task... > PlatformIO: Build` or via the `cmd-alt-b / ctrl-alt-b` hotkeys:
```
Comment (ThomasGravekamp) [inline guides/platformio.md:91]: Again a space between "firmware" and ":". You may consider making this a list.

**Review by ThomasGravekamp — CHANGES_REQUESTED** [review]
Some minor things. We can merge after you've fixed these.

**Review by ThomasGravekamp — APPROVED** [review]

**Pull request merged**